tamagotchi1/src/lib.rs

```diff
@@ -1,81 +1,207 @@
+//! #3. Homework "Tamagotchi"
 #![no_std]
-//https://academy.gear.rs/inner/hello-world-advanced/more-advanced/
-use gstd::{msg, prelude::*, debug}; //ERROR
-use tamagotchi_io::InputMessages;
-
-static mut GREETING: Option<String> = None;
-
-// #[derive(Encode, Decode, TypeInfo)]
-// pub enum InputMessages {
-//     SendHelloTo(ActorId),
-//     SendHelloReply,
-// }
-
-// #[no_mangle]
-// extern "C" fn handle() {
-//     msg::reply(String::from("Hello"), 0)
-//         .expect("Error in sending a reply message");
-// }
-//
-// #[no_mangle]
-// extern "C" fn init() {
-//     // let init_message: String = msg::load()
-//     //     .expect("Can't load init message");
-//     // debug!("Program was initialized with message {:?}",
-//     //     init_message);
-// }
+
+use gstd::{msg, debug, prelude::*, exec::{block_timestamp, block_height}};
+use tamagotchi_io::{Tamagotchi, TmgAction, TmgEvent};
+
+static mut STATE: Option<Tamagotchi> = None;
+static mut MOOD: Option<Mood> = None;
+
+
+//Mood: Fed (from 1 to 10000), Happy (from 1 to 10000) and Rested (from 1 to 10000).
+pub struct Mood {
+    pub fed: u32,
+    pub happy: u32,
+    pub rested: u32,
+}
+
+const INT_NUM: u32 = 500;
+const MAX_NUM: u32 = 10000;
+
+impl Mood {
+    const HUNGER_PER_BLOCK: u32 = 1;
+    const ENERGY_PER_BLOCK: u32 = 2;
+    const BOREDOM_PER_BLOCK: u32 = 2;
+
+    const FILL_PER_FEED: u32 = 1000;
+    const FILL_PER_ENTERTAINMENT: u32 = 1000;
+    const FILL_PER_SLEEP: u32 = 1000;
+
+
+    fn feed(&mut self) {
+        let tamagotchi = unsafe {
+            STATE.as_mut().expect("The contract is not initialized")
+        };
+
+        assert_eq!(msg::source(), tamagotchi.owner, "Only owner can feed the tamagotchi");
+        assert!(self.fed < 5000, "Tamagotchi not enough hungry");
+
+
+        // calculating and normalizing hunger level
+        let lost = (block_height() - tamagotchi.fed_block) * Self::HUNGER_PER_BLOCK;
+
+        self.fed += Self::FILL_PER_FEED;
+
+        self.fed = match lost {
+            hl if self.fed < hl => 0,
+            hl if MAX_NUM < hl => MAX_NUM,
+            _ => self.fed - lost,
+        };
```

Comment:
You can use `min`/`max` here. Also, we are to compare `self.fed` with `MAX_NUM`.
```suggestion
        self.fed -= lost.min(self.fed);
        self.fed = self.fed.min(MAX_NUM);
```